rfd/0093-offline-access.md

```diff
@@ -0,0 +1,190 @@
+metadata:
+  name: ssh-strict
+spec:
+  options:
+    record_session:
+      ssh: strict
+```
+
+Strict mode, which is only available when recording at the node, will result in the session failing to launch if the node is unable to record the session. In the event that 
+
+
+#### Moderated Sessions 
+
+Tracking participants for a moderated session requires persisting the `types.SessionTracker` resource in the backend so that the joining criteria may be evaluated when a user creates or joins a session. In the event that persisting the `types.SessionTracker` fails, the session must be aborted to prevent any unauthorized access. This is true for both proxy and node recording mode. 
+
+#### Web UI
+
+All access via the Web UI requires a connection to Auth. None of the auth connector resources are replicated to the Proxy which means users are unable to login to the Web UI without
+connectivity to Auth. Even when already logged in prior to losing connection to Auth, the Proxy doesn't perform any RBAC for users locally. It instead relies on proxying all requests
+to the Auth server on behalf of the logged in user for RBAC to be performed. 
+
+### Auth Connection Not Required 
+
+### tsh ssh
+
+Attempting to connect directly to a node via hostname, IP Address, or UUID should not require a connection to Auth for any configurations except those mentioned above.
+
+```bash
+tsh ssh foo@bar
+tsh ssh foo@127.0.0.1
+tsh ssh foo@f409a8b6-1275-4f0f-887e-8185426147f4
+```
+
+However, this doesn't work to date for two main reasons:
+
+1) `proto.AuthService/IsMFARequired`
+2) `types.SessionTracker`
+
+
+###### Per-Session MFA
+`tsh` determines whether per-session MFA is required **prior** to connecting to a node. Which means that every invocation of `tsh ssh` requires a connection to the Auth server. 
+
+Successful connection attempt when Auth is available:
+```mermaid
+sequenceDiagram
+    participant tsh
+    participant Auth
+    participant Node
+
+    tsh->>+Auth: IsMFARequired
+    Auth-->>-tsh: yes
+    tsh->>+Auth: perform mfa ceremony
+    Auth-->>+tsh: issue challenge
+    tsh-->>+Auth: challenge response
+    Auth-->>+tsh: issue certificates
+    tsh->>+Node: Connect
+    Node-->>+tsh: Success
+```
+
+Failed connection attempt when Auth is unavailable:
+```mermaid
+sequenceDiagram
+    participant tsh
+    participant Auth
+    participant Node
+
+    tsh->>+Auth: IsMFARequired
+    Auth-->>-tsh: Error
+```
+
+However, the decision if the connection should be established is ultimately made by the node, not the Auth server. The node
+performs an RBAC check on all inbound connections to determines if among other things per-session MFA is required and the MFA ceremony was 
+performed. Thus, `tsh` should **always** attempt to connect to a node even if the `proto.AuthService/IsMFARequired` RPC fails. In the worst
+case this makes `tsh ssh` for per-session MFA enabled connections take longer to fail. But that cost comes at the ability to allow all
```

Comment:
There are several scenarios where the connection attempt "takes longer to fail" if this feature is implemented. What are the expected timings in this case? "Fail quickly" is a usability feature (avoid wasting time on waiting); how can we ensure this is kept minimal?

Assuming the successful path: can we *decrease* connection latency by making the connection to auth server optional and executed in an async fashion, and only blocking if truly needed?

Reply:
Nice explanation and graphs  👍